parlai/agents/transformer/modules/decoder.py

```diff
@@ -451,3 +234,463 @@ def _apply_model_parallel(self, tensor, encoder_output, encoder_mask, incr_state
         }
 
         return tensor_out, new_incr_state
+
+
+class BaseTransformerDecoderLayer(nn.Module, ABC):
+    def __init__(
+        self,
+        opt: Opt,
+        n_heads: int = None,
+        embedding_size: int = None,
+        ffn_size: int = None,
+        attention_dropout: float = 0.0,
+        relu_dropout: float = 0.0,
+        dropout: float = 0.0,
+        activation: str = 'relu',
+        variant: str = 'aiayn',
+        **kwargs,
+    ):
+        super().__init__()
+
+        n_heads = default(n_heads, opt['n_heads'])
+        embedding_size = default(embedding_size, opt['embedding_size'])
+        ffn_size = default(ffn_size, opt['ffn_size'])
+
+        self.opt = opt
+        self.dim = embedding_size
+        self.ffn_dim = ffn_size
+        self.variant = variant
+        self.activation = activation
+        self.dropout = nn.Dropout(p=dropout)
+
+        self.self_attention = self.build_self_attention(
+            n_heads=n_heads, dim=embedding_size, dropout=attention_dropout
+        )
+        self.norm1 = torch.nn.LayerNorm(embedding_size, eps=LAYER_NORM_EPS)
+
+        self.ffn = self.build_feedforward(
+            dim=embedding_size,
+            dim_hidden=ffn_size,
+            relu_dropout=relu_dropout,
+            activation=activation,
+        )
+        self.norm3 = torch.nn.LayerNorm(embedding_size, eps=LAYER_NORM_EPS)
+
+    def build_self_attention(
+        self, n_heads: int = None, dim: int = None, dropout: float = 0
+    ) -> MultiHeadAttention:
+        raise NotImplementedError
+
+    def build_feedforward(
+        self,
+        dim: int = None,
+        dim_hidden: int = None,
+        relu_dropout: float = 0,
+        activation: str = 'relu',
+    ) -> TransformerFFN:
+        raise NotImplementedError
+
+    def reorder_incremental_state(
+        self, incremental_state: Dict[str, dict], inds: torch.Tensor
+    ) -> Dict[str, dict]:
+        """
+        Reorder all incremental-state tensors for this layer.
+        """
+        attn_types = {'self_attn': self.self_attention}
+        return {
+            attn_type: attn.reorder_incremental_state(
+                incremental_state[attn_type], inds
+            )
+            for attn_type, attn in attn_types.items()
+        }
+
+    def _create_selfattn_mask(self, x):
+        # figure out how many timestamps we need
+        bsz = x.size(0)
+        time = x.size(1)
+        # make sure that we don't look into the future
+        mask = torch.tril(x.new(time, time).fill_(1))
+        # broadcast across batch
+        mask = mask.unsqueeze(0).expand(bsz, -1, -1)
+        return mask
+
+
+###########################
+# ENCODER-DECODER MODULES #
+###########################
+
+
+@swappable(
+    self_attention=MultiHeadAttention,
+    encoder_attention=MultiHeadAttention,
+    feedforward=TransformerFFN,
+)
+class TransformerDecoderLayer(BaseTransformerDecoderLayer):
```

Comment:
i dont think this class implements building self attention; is that necessary?

also, why is that function required to be implemented in the first place?

Reply:
Good catch! Will update.

Reply:
> also, why is that function required to be implemented in the first place?

The way `@swappable` is currently implemented, you can't have subclass where the parent and child are both swappable.

So `BaseTransformerDecoderLayer` doesn't have access to `self.swappables.feedforward`, and therefore these swappable modules need to be instantiated in the concrete `DecoderLayer` classes.

Reply:
cool, could you possibly leave a comment (if you haven't already) explaining that in the code?

Reply:
Added some comments to hopefully make it clearer, and went ahead and added a default implementation to `BaseTransformerDecoderLayer` in case someone wants to subclass it and _not_ use `@swappable`.